Fix ArrayAccess limitations

aidan-casey opened the pull request:
This PR fixes #216 and the limitations it describes.

brendt commented:
Sorry it took so long, but this one looks good to merge! Can you rebase your branch? I've merged this one that's causing conflicts: https://github.com/spatie/data-transfer-object/pull/217

aidan-casey commented:
@brendt Hmm... So it looks like that PR isn't accounting for key / values (noticing I did not account for that in the array caster of this PR either). Looks like I'm going to need to do some refactoring here, so it will probably be Friday or Saturday before I can get that done.

brendt commented:
No problem :) Thanks for the heads-up!

aidan-casey commented:
@brendt - Would appreciate a review on this. A couple highlights:

- Renamed `makeItem` to `castItem` to fit with the rest of the class.
- Added `mapInto` method which casts and maps items to their appropriate key in an `array` or `Traversable` (this deduplicates code and keeps support for key mapping in arrays).
- Reverted type hinting on `$value` from `array` to `mixed` because setting it to `array` was a breaking change. I thought about union types here, but since the entry variable is `mixed` I just kept that.
- Removed contents of the `ArrayCaster` from the README and just added a link.

If this gets merged, I recommend squashing commits because I forgot about the style fixer and got in a fight with it. 😅

brendt commented on CHANGELOG.md:
```diff
@@ -6,6 +6,10 @@ All notable changes to `data-transfer-object` will be documented in this file
 
 - Fix for an empty value being created when casting `ArrayAccess` objects (#216)
 - Add logic exception when attempting to cast `ArrayAccess` objects that are not traversable (#216)
+<<<<<<< HEAD
```
Careful, there's a merge conflict committed here

aidan-casey replied:
Yeah that was my bad (🤦🏻‍♂️), I caught it too late. Everything should be clean in the latest commit, so a squash and merge would clean up that history. Alternatively, I can clean up the Git history manually with a reset.

brendt replied:
Don't worry about the history :)

Let me know when this is ready?

brendt commented on README.md:
```diff
@@ -176,8 +176,11 @@ abstract class BaseDataTransferObject extends DataTransferObject
-Any caster can be passed custom arguments, the built-in `ArrayCaster` implementation is such an example:
+Any caster can be passed custom arguments, the built-in `ArrayCaster` [implementation](https://github.com/spatie/data-transfer-object/blob/master/src/Casters/ArrayCaster.php) is a good example of how this may be used.
 
+<<<<<<< HEAD
```
Same here

aidan-casey commented:
@brendt - I think this is good to go.

brendt commented:
Good!

brendt commented:
Sorry for the late response btw, going to tag it now.

brendt commented:
Tagged: https://github.com/spatie/data-transfer-object/releases/tag/3.4.0
